File: game-engine/physics-game/CollisionManager.js
.pragma library

// CollisionManager.js is a JS layer between the front end and the C++ collision checking
// It's meant to access object specific things (like checking if rect or circle) and use that to
//      correctly call into collisionManager.cpp

// In the future I plan to move collision checking to a worker script but for initial implementation this works and
//      shows 0 performance hits so far

function shapesCollide(shapeA, shapeB) {
    if (shapeA.colliderRectangle) {
        return rectangleWillCollide(shapeA.getTopLeftCoordinate(), shapeA.getBotRightCoordinate(), shapeB);
    } else {
        return circleWillCollide(shapeA.getCenterCoordinates(), shapeA.radius, shapeB);
    }
}


// From https://stackoverflow.com/questions/401847/circle-rectangle-collision-detection-intersection
// User e.James - https://stackoverflow.com/users/33686/e-james
function circleAndRectangleCollide(cirCenter, cirRadius, rectCenter, rectWidth, rectHeight) {
    var circleDistanceX = Math.abs(cirCenter.x - rectCenter.x);
    var circleDistanceY = Math.abs(cirCenter.y - rectCenter.y);

    if (circleDistanceX > (rectWidth/2 + cirRadius)) { return false; }
    if (circleDistanceY > (rectHeight/2 + cirRadius)) { return false; }

    if (circleDistanceX <= (rectWidth/2)) { return true; }
    if (circleDistanceY <= (rectHeight/2)) { return true; }

    var cornerDistance_sq = (circleDistanceX - rectWidth/2)^2 +
                         (circleDistanceY - rectHeight/2)^2;

    return (cornerDistance_sq <= (cirRadius^2));
}

function circleWillCollide(cirCenter, cirRadius, otherShape) {
    if (otherShape.colliderRectangle) { // Circle && Rect
        return circleAndRectangleCollide(cirCenter, cirRadius, otherShape.getCenterCoordinates(), otherShape.width, otherShape.height);
    } else { // Circle && Circle
        return collisionManager.circlesCollide(cirCenter, cirRadius, otherShape.getCenterCoordinates(), otherShape.radius);
    }
}

function rectangleWillCollide(rectPoint1, rectPoint2, otherShape) {
    if (otherShape.colliderRectangle) { // Rect && Rect
        return collisionManager.rectanglesCollide(rectPoint1, rectPoint2, otherShape.getTopLeftCoordinates(), otherShape.getBotRightCoordinates());
    } else { // Rect && Circle
        return circleAndRectangleCollide(otherShape.getCenterCoordinates(), otherShape.radius, getRectangleCenterCoordinates(rectPoint1, rectPoint2), Math.abs(rectPoint1.x - rectPoint2.x), Math.abs(rectPoint1.y - rectPoint1.y));
    }
}

function getRectangleCenterCoordinates(rectPoint1, rectPoint2) {
    return Qt.point((rectPoint1.x + rectPoint2.x) / 2, (rectPoint1.y + rectPoint1.y) / 2);
}

File: game-engine/physics-game/collisionmanager.cpp
#include "collisionmanager.h"
#include <QDebug>
#include <QVariantMap>
#include <QtMath>

CollisionManager::CollisionManager(QObject *parent) : QObject(parent)
{

}

bool CollisionManager::collision(QQuickItem *a, QQuickItem *b)
{
//    qDebug() << "CollisionManager::collision" << a->width();
    const CircleCollider *cir = qobject_cast<const CircleCollider *>(a);
    if (cir != NULL) {
        const CircleCollider *cir2 = qobject_cast<const CircleCollider *>(b);
        if (cir2 != NULL) {
//            qDebug() << "Testing circles";
            return circlesCollide(cir, cir2);
        }

        const RectangleCollider *rect2 = qobject_cast<const RectangleCollider *>(b);
        if (rect2 != NULL) {
            return circleAndRectangleCollide(cir, rect2);
        }
        return false;
    }

    const RectangleCollider *rect = qobject_cast<const RectangleCollider *>(a);
    if (rect != NULL) {
        const CircleCollider *cir2 = qobject_cast<const CircleCollider *>(b);
        if (cir2 != NULL) {
            return circleAndRectangleCollide(cir2, rect);
        }

        const RectangleCollider *rect2 = qobject_cast<const RectangleCollider *>(b);
        if (rect2 != NULL) {
            return rectanglesCollide(rect, rect2);
        }
        return false;
    }

    return false;
}

bool CollisionManager::rectanglesCollide(const RectangleCollider *a, const RectangleCollider *b)
{
    return rectanglesCollide(a->getTopLeftCoordinates(), a->getBotRightCoordinates(), b->getTopLeftCoordinates(), b->getBotRightCoordinates());
}

bool CollisionManager::rectanglesCollide(const QPoint topLeftA, const QPoint botRightA, const QPoint topLeftB, const QPoint botRightB)
{
    return (topLeftA.x() < botRightB.x() &&
             topLeftB.x() < botRightA.x() &&
             topLeftA.y() < botRightB.y() &&
             topLeftB.y() < botRightA.y());
}

bool CollisionManager::circlesCollide(const CircleCollider *a, const CircleCollider *b)
{
    return circlesCollide(a->getCenterCoordinates(), a->getRadius(), b->getCenterCoordinates(), b->getRadius());
}

bool CollisionManager::circlesCollide(const QPoint cirACenter, int aRadius, const QPoint cirBCenter, int bRadius)
{
    return distanceBetweenCircles(cirACenter, cirBCenter) < aRadius + bRadius;
}

int CollisionManager::distanceBetweenCircles(const QPoint aCenter, const QPoint bPoint)
{
    int dx = aCenter.x() - bPoint.x();
    int dy = aCenter.y() - bPoint.y();
    return qSqrt(dx * dx + dy * dy);
}

bool CollisionManager::circleAndRectangleCollide(const CircleCollider *cir, const RectangleCollider *rect)
{
    return circleAndRectangleCollide(cir->getCenterCoordinates(), cir->getRadius(), rect->getCenterCoordinates(), rect->width(), rect->height());
}

// From https://stackoverflow.com/questions/401847/circle-rectangle-collision-detection-intersection
// User e.James - https://stackoverflow.com/users/33686/e-james
bool CollisionManager::circleAndRectangleCollide(const QPoint cirCenter, int cirRadius, const QPoint rectCenter, int rectWidth, int rectHeight)
{
    int circleDistanceX = qFabs(cirCenter.x() - rectCenter.x());
    int circleDistanceY = qFabs(cirCenter.y() - rectCenter.y());

    int halfRectWidth = rectWidth;//rectWidth / 2;
    int halfRectHeight = rectHeight;//rectHeight / 2;

    if (circleDistanceX > (halfRectWidth + cirRadius) || circleDistanceY > (halfRectHeight + cirRadius))
    {
        return false;
    }

    if (circleDistanceX <= (halfRectWidth) || circleDistanceY <= (halfRectHeight))
    {
        return true;
    }

    return (((circleDistanceX - halfRectWidth) ^ 2) + ((circleDistanceY - halfRectHeight) ^ 2) <= (cirRadius ^ 2));
}


